### ISSUE TITLE
Querying on modified_since/pushed_since is very slow

### ISSUE THREAD

**Issue by owtaylor** [fedora-infra/bodhi#3064]
To incrementally query for recently changed Bodhi updates, you need to do three queries:

```
  /updates?submitted_since=[timestamp]&[other_params]
  /updates?modified_since=[timestamp>][other_params]
  /updates?pushed_since=[timestamp]&[other_params]
```

(might be nice to have a "changed_since" that combines the three with an OR). The submitted_since query is typically fast, because that column is indexed, but the other columns aren't indexed, and the query ranges between slow and very slow depending on other_params. E.g. if you do

```
 /updates?modified_since=timestamp&content_type=rpm
```

you get a super-slow query because the database (I'm guessing) finds all rpm builds, iterates through them and looks up the timestamp for each one.

[ I'm using this to update a status page for Fedora flatpaks - https://fedora.fishsoup.net/flatpak-status/ - it correlates Flatpak updates with the update status for all the individual RPMs within the Flatpak. Always querying Bodhi for all the updates would be prohibitive - it takes 30 minutes or more of banging on the Bodhi API  ]

**Comment by bowlofeggs** [comment]
Do you think Bodhi's RSS feeds could be useful for what you are trying to do?

**Comment by owtaylor** [comment]
> Do you think Bodhi's RSS feeds could be useful for what you are trying to do?

As far as I can see, /rss/updates is just an alternate formatting of /updates - so it's not obviously helpful to me - it omits some information I don't need like the full comment stream, but is also missing information I do need - like the full list of builds.

**Comment by sebwoj** [comment]
Running the following script (quote insetad of "insert code" because of formatting issues):

> #!/bin/bash
url=https://bodhi.fedoraproject.org/updates/
printf "\n----==== modified_since ====----\n"
time curl -G -o m.out --data "modified_since=2018-03-20&content_type=rpm" $url
printf "\n----==== submitted_since ====----\n"
time curl -G -o s.out --data "submitted_since=2018-03-20&content_type=rpm" $url
printf "\n----==== pushed_since ====----\n"
time curl -G -o p.out --data "pushed_since=2018-03-20&content_type=rpm" $url`

Results with the following output:

>----==== modified_since ====----
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  135k  100  135k    0     0  11702      0  0:00:11  0:00:11 --:--:-- 19995
real	0m11.832s
user	0m0.025s
sys	0m0.013s
----==== submitted_since ====----
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 54650  100 54650    0     0   5807      0  0:00:09  0:00:09 --:--:--  7974
real	0m9.425s
user	0m0.025s
sys	0m0.011s
----==== pushed_since ====----
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 67312  100 67312    0     0   5015      0  0:00:13  0:00:13 --:--:--  7807
real	0m13.435s
user	0m0.030s
sys	0m0.009s

@owtaylor What query parameters do you use when you run your query?

**Comment by owtaylor** [comment]
Thanks for investigating this!

I'm not reproducing the same numbers as I did when I submitted the bug - I think this is largely a combination of different data in the database, and different database load.

I'm still seeing a strong difference with adding &content_type=rpm ... if you remove that, your script speeds up by a factor of 4. This indicates to me that the query is potentially optimizable at the database level - *but* I'm no longer seeing the difference between submitted_since and modified_since I did before, which makes it less clear that adding indexes on modified_since/pushed_since would actually help substantially. It may have been testing error, or it may be that the DB optimizer is choosing different query plans today.

The original query I was dealing with was more like:

 ?rows_per_page=100&modified_since=2018-03-20&content_type=rpm&active_releases=1

That's slower, but possibly only because it's retrieving more data.

**Comment by sebwoj** [comment]
Duplication #271

**Comment by bowlofeggs** [comment]
I wouldn't quite call it a duplicate since this one is specific to some particular query parameters, but it's fair to say it's related for sure. I think I'll opt to keep this open due to that.

**Comment by mattiaverga** [comment]
Fixed by #4048  please check
